graphql_compiler/deserialization.py

```diff
@@ -0,0 +1,253 @@
+        # represented as strings.
+        GraphQLInt.name: int,
+        # Bypass the GraphQLBoolean parser and allow some strings and ints as input.
+        GraphQLBoolean.name: _custom_boolean_deserialization,
+    }
+)
+
+_JSON_TYPES_AND_DESERIALIZATION_FUNCTIONS: Mapping[
+    str, Tuple[Tuple[Type, ...], Callable[[Any], Any]]
+] = MappingProxyType(
+    {
+        scalar_type.name: (
+            _ALLOWED_JSON_SCALAR_TYPES[scalar_type.name],
+            _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(scalar_type.name, scalar_type.parse_value),
+        )
+        for scalar_type in SUPPORTED_SCALAR_TYPES
+    }
+)
+
+
+def convert_scalar_value_to_graphql_type(expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Convert a scalar value to the appropriate type for the given GraphQLScalarType.
+
+    Args:
+        expected_type: a GraphQLScalarType to which value should be converted.
+        value: object that can be interpreted as being of expected_type.
+
+    Returns:
+        a value of the type produced by the parser of the expected type:
+            GraphQLDate: datetime.date
+            GraphQLDateTime: datetime.datetime with tzinfo=pytz.utc
+            GraphQLFloat: float
+            GraphQLDecimal: decimal.Decimal
+            GraphQLInt: int
+            GraphQLString: str
+            GraphQLBoolean: bool
+            GraphQLID: str
+
+    Raises:
+        TypeError: if the value was not of the expected type.
+    """
+    types_and_deserialization = _JSON_TYPES_AND_DESERIALIZATION_FUNCTIONS.get(expected_type.name)
+    if types_and_deserialization is None:
+        raise TypeError(
+            f"Unexpected GraphQLType {expected_type}. No deserialization function known."
+        )
+    expected_python_types, deserialization_function = types_and_deserialization
+    if not isinstance(value, expected_python_types):
+        raise TypeError(
+            f"{value} ({type(value)} cannot be deserialized to GraphQL type {expected_type}."
+        )
+    return deserialization_function(value)
+
+
+def deserialize_scalar_argument(name: str, expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Deserialize a serialized scalar argument. See docstring of deserialize_argument.
+
+    Args:
+        name: the name of the argument.
+        expected_type: a GraphQLScalarType to which value should be converted.
+        value: object that can be interpreted as being of expected_type.
+
+    Returns:
+        a value of the type produced by the parser of the expected type:
+            GraphQLDate: datetime.date
+            GraphQLDateTime: datetime.datetime with tzinfo=pytz.utc
+            GraphQLFloat: float
+            GraphQLDecimal: decimal.Decimal
+            GraphQLInt: int
+            GraphQLString: str
+            GraphQLBoolean: bool
+            GraphQLID: str
+
+    Raises:
+        GraphQLInvalidArgumentError: if the argument value was not of the expected type.
```

Comment:
4x docstring repetition is a bit much. If you agree that we should make this function private, do you mind making this docstring a one-liner?